Implement partial fourier basis (fixes #27)

scottgigante opened the pull request:
Implements partial computation of the fourier basis for plotting of the Laplacian Eigenmap and/or spectral clustering without the need for full eigendecomposition. API discussed in #27.

nperraud commented on pygsp/graphs/fourier.py:
```diff
@@ -88,26 +103,43 @@ def compute_fourier_basis(self, recompute=False):
 
         assert self.L.shape == (self.N, self.N)
-        if self.N > 3000:
+        if self.N > 3000 and n_eigenvectors == self.N:
```
Shall we change this test to something like: self.N*n_eigenvectors > 3000**2 ?

mdeff replied:
Agreed. It should ideally represent the true computational cost. I know it's O(N**3) for the full eigendecomposition. Do you know for the partial? I.e. dependency on matrix size and number of eigenvectors.

scottgigante replied:
I've had a look and estimating the complexity of `scipy.sparse.linalg.eigsh` (which uses the Lanczos method from ARPACK) is not trivial. My best guess is to use N^2 k, though this is not based on any solid knowledge.

mdeff replied:
Ok. I don't know either. I agree on N^2 k as a rough estimate.

nperraud commented on pygsp/tests/test_graphs.py:
```diff
@@ -75,8 +75,21 @@ def test_fourier_basis(self):
         assert G.e[-1] < 2
+        assert np.testing.assert_allclose(U, G.U[:, -n:])
+        assert np.testing.assert_allclose(e, G.e[-n:])
 
```
shall we also test the function set_coordinates?

nperraud commented:
Thanks a lot for the PR. I left two small comments. I think you did a good job, but I guess @mdeff should be the one to do the merging.

nperraud commented:
The test are failing with errors like this:
```
File "pygsp/utils.py", line 255, in utils.py
Failed example:
    utils.symmetrize(W, method='tril')
Expected:
    array([[0., 3., 4.],
           [3., 1., 2.],
           [4., 2., 3.]])
Got:
    array([[ 0.,  3.,  4.],
           [ 3.,  1.,  2.],
           [ 4.,  2.,  3.]])
```
Any idea what is going on?

scottgigante commented:
Both valid comments. I've given it a slightly different warning message depending on whether the user asks for full eigendecomposition or partial with many eigenvectors, but if you have a preferred message I can concatenate them.

The tests thing is weird - I know `doctest` does require string equivalence of test results, but I didn't change `utils` and I think default behaviour shouldn't have changed. Looks like a `numpy` or Python version change, to me.

scottgigante commented:
Just an update on those failing tests: one of them was actually my test, hidden among the others. Should be fixed now. The others are currently also failing on `master`: https://travis-ci.org/epfl-lts2/pygsp/jobs/382958006

scottgigante commented:
Actually it still doesn't pass:
```
>>> np.testing.assert_allclose(np.abs(U), np.abs(G.U[:,-n:]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/scottgigante/.local/lib/python2.7/site-packages/numpy/testing/utils.py", line 1395, in assert_allclose
    verbose=verbose, header=header, equal_nan=equal_nan)
  File "/home/scottgigante/.local/lib/python2.7/site-packages/numpy/testing/utils.py", line 778, in assert_array_compare
    raise AssertionError(msg)
AssertionError:
Not equal to tolerance rtol=1e-07, atol=0

(mismatch 0.0722061242071%)
 x: array([[  2.714909e-02,   1.355861e-02,   5.760985e-02, ...,
          2.945387e-08,   1.336955e-08,   2.885479e-08],
       [  3.213260e-02,   1.889188e-02,   5.686249e-02, ...,...
 y: array([[  2.714909e-02,   1.355861e-02,   5.760985e-02, ...,
          2.945387e-08,   1.336955e-08,   2.885479e-08],
       [  3.213260e-02,   1.889188e-02,   5.686249e-02, ...,...
>>> np.testing.assert_allclose(np.abs(U), np.abs(G.U[:,-n:]), atol=1e-13)
>>> np.testing.assert_allclose(np.abs(U), np.abs(G.U[:,-n:]), rtol=1e-2)
```
Either of setting absolute tolerance or relative tolerance a little higher works. Absolute tolerance makes more sense to me but I'm open to suggestions here.

mdeff commented on pygsp/graphs/fourier.py:
```diff
@@ -43,14 +44,18 @@ def mu(self):
+        n_eigenvectors : int or `None`
+            Number of eigenvectors to compute. If `None`, all eigenvectors
+            are computed. (default: None)
+
```
No empty line between parameters.

mdeff commented on pygsp/graphs/fourier.py:
```diff
@@ -88,26 +103,49 @@ def compute_fourier_basis(self, recompute=False):
 
-        if hasattr(self, '_e') and hasattr(self, '_U') and not recompute:
+        if hasattr(self, '_e') and hasattr(self, '_U') and not recompute \
+                and n_eigenvectors <= len(self.e):
```
I'm not a fan of the `\` notation for line continuation. Can we use parentheses?

mdeff commented on pygsp/graphs/fourier.py:
```diff
@@ -88,26 +103,49 @@ def compute_fourier_basis(self, recompute=False):
+        else:
+            # fast partial eigendecomposition of hermitian matrices
+            # this will be slow if L is non-symmetric
+            if np.sum(self.L != self.L.T) != 0:
```
I don't think we should test for symmetry. It's assumed (in the current implementation at least) that `L` is symmetric. We only support the combinatorial and normalized Laplacian for now. If at some point we had support for non-symmetric laplacians, we'll have to change many things (and we'll test the `lap_type`). As we're using the hermitian version of the eigensolvers, only the lower or upper triangular part of the matrix will be considered anyway.

mdeff commented on pygsp/graphs/graph.py:
```diff
@@ -187,7 +187,8 @@ def set_coordinates(self, kind='spring', **kwargs):
             or Nx3 to set the coordinates manually or the name of a layout
             algorithm. Available algorithms: community2D, random2D, random3D,
-            ring2D, line1D, spring. Default is 'spring'.
+            ring2D, line1D, spring, laplacian_eigenmap2D, laplacian_eigenmap3D.
```
Would be nice to have an `n_dimensions` parameter here instead of `laplacian_eigenmap2D`, `laplacian_eigenmap3D`. User can then do an eigenmap in n dimensions. Should also convert the others. But I won't object here, as this issue predates that PR.

mdeff replied:
Sure, no prob.

mdeff commented:
I've fixed the issues (361f67e, 57f370a) in master. You can rebase for the tests to pass.

scottgigante commented:
Thanks for the review @mdeff ! I've implemented those changes and rebased from master so the tests should pass, I think. Re your comment: "give a potential solution: do a partial decomposition, or use polynomial filters if the goal is filtering. Maybe mention the filtering of random signals as well." - I'm not quite sure what I would say about filtering random signals. If you have better suggested wording than the current warning I'm happy to put it in.

scottgigante commented:
For reference, the decrease in coverage is due to the slow eigendecomposition warning not being tested. We could add a simple large problem (eigendecomposition of 3000x3000 identity?) if this is necessary, but I don't think we need to test that warning.

mdeff commented on pygsp/graphs/fourier.py:
```diff
@@ -88,26 +102,38 @@ def compute_fourier_basis(self, recompute=False):
+            # filter's implementations may not anticipate. Better for plotting
+            # too.
+            assert -1e-12 < self._e[0] < 1e-12
+            self._e[0] = 0
```
Can we do those tests after the if-else branch?

scottgigante replied:
A partial eigendecomposition won't compute the smallest eigenvector, so there's no reason why the smallest eigenvector _would_ be 0.

mdeff commented on pygsp/graphs/fourier.py:
```diff
@@ -88,26 +102,38 @@ def compute_fourier_basis(self, recompute=False):
+            self.logger.warning(
+                'Computing the {0} eigendecomposition of a large matrix ({1} x'
+                ' {1}) is expensive. Consider decreasing n_eigenvectors '
+                'or, if using the Fourier basis as a filter, using a '
```
as a filter => to filter

scottgigante commented:
Am I right in saying that partial eigendecomposition wouldn't give the smallest eigenvector? Then those checks will fail if n_eigenvectors < N

mdeff commented:
We **want** to compute the eigenvectors associated with the smallest eigenvalues. For example for spectral clustering, we often use the first `k`. For spectral embedding, we use 2 or 3. So you need to pass `which='SM'` (smallest magnitude) to `eigsh`. Moreover, in `set_coordinates`, we have to discard the constant eigenvector (index 0) and use the first 2-3.

scottgigante commented:
Ah, good point. Should we allow the user to pass `which` with `'SM'` as default?

nperraud commented:
My feeling is that for the coordinates, the user always wants the smallest ones. So I am not sure that an option makes sense here.

scottgigante commented:
For coordinates, I agree - but for `compute_fourier_basis(n_eigenvectors=k)` in general, would the user ever want the eigenvectors associated with the largest eigenvectors?

mdeff commented:
I don't see any use for the largest ones.

scottgigante commented:
Done. One test is failing on Python2.7 but I don't think it's my fault.

mdeff commented:
Not your fault, that was a transient fail. Looks good to me now. Thanks for the PR, and don't hesitate to come back for another one! :)

scottgigante commented:
Thanks!
